modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+    private final Function<String, HttpClient.PasswordAuthenticationHolder> credentialsBuilder;
+
+    EnterpriseGeoIpDownloader(
+        Client client,
+        HttpClient httpClient,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        long id,
+        String type,
+        String action,
+        String description,
+        TaskId parentTask,
+        Map<String, String> headers,
+        Supplier<TimeValue> pollIntervalSupplier,
+        Function<String, HttpClient.PasswordAuthenticationHolder> credentialsBuilder
+    ) {
+        super(id, type, action, description, parentTask, headers);
+        this.client = client;
+        this.httpClient = httpClient;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.pollIntervalSupplier = pollIntervalSupplier;
+        this.credentialsBuilder = credentialsBuilder;
+    }
+
+    void setState(EnterpriseGeoIpTaskState state) {
+        // this is for injecting the state in GeoIpDownloaderTaskExecutor#nodeOperation just after the task instance has been created
+        // by the PersistentTasksNodeService -- since the GeoIpDownloader is newly created, the state will be null, and the passed-in
+        // state cannot be null
+        assert this.state == null;
+        assert state != null;
+        this.state = state;
+    }
+
+    // visible for testing
+    void updateDatabases() throws IOException {
+        var clusterState = clusterService.state();
+        var geoipIndex = clusterState.getMetadata().getIndicesLookup().get(EnterpriseGeoIpDownloader.DATABASES_INDEX);
+        if (geoipIndex != null) {
+            logger.trace("The {} index is not null", EnterpriseGeoIpDownloader.DATABASES_INDEX);
+            if (clusterState.getRoutingTable().index(geoipIndex.getWriteIndex()).allPrimaryShardsActive() == false) {
+                throw new ElasticsearchException("not all primary shards of [" + DATABASES_INDEX + "] index are active");
+            }
+            var blockException = clusterState.blocks().indexBlockedException(ClusterBlockLevel.WRITE, geoipIndex.getWriteIndex().getName());
+            if (blockException != null) {
+                throw blockException;
+            }
+        }
+
+        logger.trace("Updating geoip databases");
+        IngestGeoIpMetadata geoIpMeta = clusterState.metadata().custom(IngestGeoIpMetadata.TYPE, IngestGeoIpMetadata.EMPTY);
+
+        // if there are entries in the cs that aren't in the persistent task state,
+        // then download those (only)
+        // ---
+        // if there are in the persistent task state, that aren't in the cluster state
+        // then nuke those (only)
+        // ---
+        // else, just download everything
+        boolean addedSomething = false;
+        {
+            EnterpriseGeoIpTaskState _state = state;
+            Set<String> metas = Set.copyOf(_state.getDatabases().keySet());
```

Comment:
This seems odd to me to copy this only to use it once? Is there a reason you did this? I don't think we have it for consistency since you only used it once in this block?

Reply:
I'm going to leave this one for @joegallo to comment on. Maybe he's avoiding some race condition since this method is running in the cluster state applier thread, and something else could be modifying the task state at the same time?

Reply:
I'll drop it -- the underlying map is immutable anyway, so it's not like we need to worry about the keyset mutating out from underneath us. I imagine it was just me thinking defensively for a little bit there.

Reply:
Dropped in 39416f0d751.